executor/aggregate.go

```diff
@@ -395,32 +397,49 @@ func (w *HashAggPartialWorker) shuffleIntermData(sc *stmtctx.StatementContext, f
 }
 
 // getGroupKey evaluates the group items and args of aggregate functions.
-func (w *HashAggPartialWorker) getGroupKey(sc *stmtctx.StatementContext, row chunk.Row) ([]byte, error) {
-	w.groupValDatums = w.groupValDatums[:0]
-	for _, item := range w.groupByItems {
-		v, err := item.Eval(row)
+func getGroupKey(ctx sessionctx.Context, input *chunk.Chunk, groupKey [][]byte, groupByItems []expression.Expression) ([][]byte, error) {
+	numRows := input.NumRows()
+	avlGroupKeyLen := mathutil.Min(len(groupKey), numRows)
+	for i := 0; i < avlGroupKeyLen; i++ {
+		groupKey[i] = groupKey[i][:0]
+	}
+	for i := avlGroupKeyLen; i < numRows; i++ {
+		groupKey = append(groupKey, make([]byte, 0, 10*len(groupByItems)))
+	}
+
+	for _, item := range groupByItems {
+		tp := item.GetType()
+		buf, err := expression.GetColumn(tp.EvalType(), numRows)
 		if err != nil {
 			return nil, err
 		}
-		// This check is used to avoid error during the execution of `EncodeDecimal`.
-		if item.GetType().Tp == mysql.TypeNewDecimal {
-			v.SetLength(0)
+
+		if groupKey, err = expression.VectorizedGetGroupKey(ctx, groupKey, item, tp, input, buf); err != nil {
```

Comment:
We should recycle `buf` here.

Reply:
See the next line. I put the ```buf``` back into our pool at the end of the loop.